sample-app/src/components/InputDropdown.tsx

```diff
@@ -52,74 +47,120 @@ export default function InputDropdown({
+    }
+    
+    const modifiedOnClickOption : typeof child.props.onClickOption = (option: typeof child.props.options[0]) => {
+      child.props.onClickOption(option);
+      dispatch({ type: 'HideSections' });
+    }
+
+    const modifiedOnFocusChange = (value: string, focusedOptionId: string) => {
+      child.props.onFocusChange(value);
+      setFocusedOptionId(focusedOptionId);
+    };
+
+    if (focusedSectionIndex === undefined) {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'reset', onClickOption: modifiedOnClickOption });
+    } else if (index === focusedSectionIndex) {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'active', onFocusChange: modifiedOnFocusChange, onClickOption: modifiedOnClickOption });
+    } else {
+      return React.cloneElement(child, { onLeaveSectionFocus, focusStatus: 'inactive', onClickOption: modifiedOnClickOption });
+    }
+  });
+
+  const numSections = childrenWithProps.length;
+
+  function onLeaveSectionFocus(focusNext: boolean) {
+    if (focusedSectionIndex === undefined && focusNext) {
+      dispatch({ type: 'FocusSection', newIndex: 0 });
+    } else if (focusedSectionIndex !== undefined && focusedSectionIndex < numSections - 1) {
+      let newSectionIndex: number | undefined = focusNext ? focusedSectionIndex + 1 : focusedSectionIndex - 1;
+      if (newSectionIndex < 0) {
+        newSectionIndex = undefined;
+        onInputChange(latestUserInput);
+      }
+      dispatch({ type: 'FocusSection', newIndex: newSectionIndex });
+    } else if (focusedSectionIndex === numSections - 1 && !focusNext) {
+      let newSectionIndex: number | undefined = focusedSectionIndex - 1;
+      if (newSectionIndex < 0) {
+        newSectionIndex = undefined;
+      }
+      dispatch({ type: 'FocusSection', newIndex: newSectionIndex });
+    }
+  }
+
   function handleDocumentClick(evt: MouseEvent) {
     const target = evt.target as HTMLElement;
-    if (!target.isSameNode(inputRef.current)) {
-      dispatch({ type: 'HideOptions' });
+    if (!(target.isSameNode(inputRef.current) || (dropdownRef.current !== null && dropdownRef.current.contains(target)))) {
+      dispatch({ type: 'HideSections' });
+    }
+  }
+
+  function handleDocumentKeydown(evt: globalThis.KeyboardEvent) {
+    if ((evt.key === 'Escape' || evt.key === 'Tab')) {
+      dispatch({ type: 'HideSections' });
     }
   }
 
   useEffect(() => {
-    document.addEventListener('click', handleDocumentClick)
-    return () => document.removeEventListener('click', handleDocumentClick);
+    if (shouldDisplayDropdown) {
+      document.addEventListener('click', handleDocumentClick);
+      document.addEventListener('keydown', handleDocumentKeydown);
+      return () => {
+        document.removeEventListener('click', handleDocumentClick);
+        document.removeEventListener('keydown', handleDocumentKeydown);
+      }
+    }
   });
 
   function onKeyDown(evt: KeyboardEvent<HTMLInputElement>) {
```

Comment:
Is this onKeyDown still needed? Can we remove this and move the Enter functionality to `handleDocumentKeydown` instead?

Reply:
`onKeyDown` tracks the keys pressed when the user has clicked into the input bar. otherwise, even if the user presses Enter for something else on the page, it would trigger the onSubmit for the input bar

Reply:
I see, we can keep the enter in onKeyDown then. But what about the arrowDown/Right in the else if? I assume the other `handleKeyDown` listeners should already take care of them no?

Reply:
yes, I moved that into `handleKeyDown`